man/man1/cdf2coord.1
.TH cdf2coord 1  "USER COMMANDS"
.SH NAME
cdf2coord \- extract CLM grid from NetCDF file
.SH SYNOPSIS
.B cdf2coord
[\-var name] [\-index i] [\-{float|double}] [\-scale s] [\-latlon] [\-raw] [\-json]
.I netcdffile gridfile
.SH DESCRIPTION
Program extracts grid file in CLM format from NetCDF input data. All cells with valid input will be put in the grid file.
.SH OPTIONS
.TP
\-var name
Variable name in NetCDF file. If not specified the first variable with more than one dimension is chosen.
.TP
\-index i
Converting from i-th dataset. Default is first (=0).
.TP
\-scale factor
Scaling factor for CLM grid file. Default is 0.01 for short values in CLM grid file, 1 for float or double values. If factor differs from 1 and -float or -double option is set a warning is printed and scaling factor reset to 1.
.TP
\-{float|double}
Write float or double values in CLM grid file without applying the scaling factor, default is short.
.TP
\-latlon
The order of how the cells are read is set to change latitude first, default is longitude first.
.TP
\-raw
Output format is raw without header, default is CLM format.
.TP
\-json
An additional JSON metafile with suffix \fI.json\fP is written.
.TP
\-json
JSON metafile is created with suffix '.json'
.TP
.I netcdffile
Filename of NetCDF file.
.TP
.I gridfile
Filename of grid data file written.
.SH EXAMPLES
.TP
Create grid file from temperature data file:
.B cdf2coord
-var tas temp1901-2010.nc grid.clm
.PP
.SH EXIT STATUS
.B cdf2coord
returns a zero exit status if the CLM file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
cdf2grid(1), cdf2clm(1), clm2cdf(1), bin2cdf(1), country2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Added

- Option `-h` added to utilities `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.
- Parsing of options corrected in `fprintincludes.c`.
- Utility `bin2cdf` fixed for timestep>1.


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ``.gitlab/merge_request_templates/default.md`` that includes several checklists.


## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.

src/utils/cdf2coord.c
/**************************************************************************************/
/**                                                                                \n**/
/**                c  d  f  2  c  o  o  r  d  .  c                                 \n**/
/**                                                                                \n**/
/**     Converts NetCDF data file into LPJ coord data file                         \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE  "Usage: %s [-var name] [-index i] [-{float|double}] [-scale s] [-latlon] [-raw] [-json] netcdffile coordfile\n"

#ifdef USE_NETCDF
#include <netcdf.h>
#endif

#define error(rc) if(rc){ fprintf(stderr,"ERROR421: Cannot read '%s': %s.\n",argv[i],nc_strerror(rc)); return EXIT_FAILURE;}

int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  int rc,ncid,var_id,*dimids,i,j,nvars,lon_id,lat_id,ndims,index,first;
  int index1,index2,len1,len2;
  double *lat,*lon;
  float scalar=0.0;
  size_t lat_len,lon_len;
  size_t offsets[4]={0,0,0,0},counts[4]={1,1,1,1};
  double missing_value,data;
  char name[NC_MAX_NAME+1],*endptr;
  Header header;
  Intcoord coord;
  struct
  {
    double lon,lat;
  } coord_d;
  struct
  {
    float lon,lat;
  } coord_f;
  Netcdf_config netcdf_config;
  char *var;
  char *out_json,*arglist;
  FILE *out;
  Bool latlon,isjson,israw,scalar_set;
  var=NULL;
  header.datatype=LPJ_SHORT;
  header.scalar=0.01;
  latlon=isjson=israw=scalar_set=FALSE;
  initsetting_netcdf(&netcdf_config);
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-var"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-var'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        var=argv[++i];
      }
      else if(!strcmp(argv[i],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[i],"-raw"))
        israw=TRUE;
      else if(!strcmp(argv[i],"-float"))
      {
        header.datatype=LPJ_FLOAT;
        header.scalar=1;
      }
      else if(!strcmp(argv[i],"-double"))
      {
        header.datatype=LPJ_DOUBLE;
        header.scalar=1;
      }
      else if(!strcmp(argv[i],"-latlon"))
        latlon=TRUE;
      else if(!strcmp(argv[i],"-index"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-index'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        offsets[0]=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for index.\n",argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-scale"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-scale'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        scalar=header.scalar=(float)strtod(argv[++i],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for scale.\n",argv[i]);
          return EXIT_FAILURE;
        }
        if(header.scalar!=1)
          scalar_set=TRUE;
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[i],argv[0]); 
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<i+2)
  {
    fprintf(stderr,"Missing argument(s).\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(header.datatype!=LPJ_SHORT && scalar_set)
  {
    fprintf(stderr,"Warning: Scaling set to %g but datatype is %s, scaling set to 1.\n",
            scalar,typenames[header.datatype]);

    header.scalar=1;
  }
  rc=nc_open(argv[i],NC_NOWRITE,&ncid);
  if(rc)
  {
    fprintf(stderr,"ERROR409: Cannot open '%s': %s.\n",
            argv[i],nc_strerror(rc));
    return EXIT_FAILURE;
  }
  if(var==NULL)
  {
    nc_inq_nvars(ncid,&nvars);
    for(j=0;j<nvars;j++)
    {
      nc_inq_varname(ncid,j,name);
      if(strcmp(name,netcdf_config.lon_bnds.name) && strcmp(name,netcdf_config.lat_bnds.name) && strcmp(name,netcdf_config.lon.name) && strcmp(name,netcdf_config.lon.standard_name) && strcmp(name,netcdf_config.lat.name) && strcmp(name,netcdf_config.lat.standard_name) && strcmp(name,netcdf_config.time.name) && strcmp(name,netcdf_config.pft.name) && strcmp(name,netcdf_config.depth.name) && strcmp(name,netcdf_config.depth_bnds.name) && strcmp(name,netcdf_config.time_bnds.name))
      {
        nc_inq_varndims(ncid,j,&ndims);
        if(ndims>1)
        {
          var_id=j;
          break;
        }
      }
    }
    if(j==nvars)
    {
      fprintf(stderr,"ERRO405: No variable found in '%s'.\n",argv[i]);
      nc_close(ncid);
      return EXIT_FAILURE;
    }
  }
  else if(nc_inq_varid(ncid,var,&var_id))
  {
    fprintf(stderr,"ERROR406: Cannot find variable '%s' in '%s'.\n",var,argv[i]);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_varndims(ncid,var_id,&ndims);
  if(ndims<2)
  {
    fprintf(stderr,"ERROR408: Invalid number of dimensions %d in '%s', must be >1.\n",
            ndims,argv[i]);
    return EXIT_FAILURE;
  }
  else if(ndims>4)
  {
    fprintf(stderr,"ERROR408: Invalid number of dimensions %d in '%s', must be <5.\n",
            ndims,argv[i]);
    return EXIT_FAILURE;
  }
  else
    first=ndims-2;
  index=ndims-1;
  dimids=newvec(int,ndims);
  if(dimids==NULL)
  {
    printallocerr("dimids");
    return EXIT_FAILURE;
  }
  nc_inq_vardimid(ncid,var_id,dimids);
  nc_inq_dimname(ncid,dimids[index],name);
  rc=nc_inq_varid(ncid,name,&lon_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[i],nc_strerror(rc));
    free(dimids);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[index],&lon_len);
  lon=newvec(double,lon_len);
  if(lon==NULL)
  {
    free(dimids);
    printallocerr("lon");
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,lon_id,lon);
  if(rc)
  {
    free(dimids);
    free(lon);
    fprintf(stderr,"ERROR410: Cannot read longitude in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    return EXIT_FAILURE;
  }
  nc_inq_dimname(ncid,dimids[index-1],name);
  rc=nc_inq_varid(ncid,name,&lat_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[i],nc_strerror(rc));
    free(dimids);
    free(lon);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[index-1],&lat_len);
  lat=newvec(double,lat_len);
  if(lat==NULL)
  {
    free(dimids);
    free(lon);
    printallocerr("lat");
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,lat_id,lat);
  if(rc)
  {
    free(dimids);
    free(lon);
    free(lat);
    fprintf(stderr,"ERROR410: Cannot read latitude in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    return EXIT_FAILURE;
  }

  rc=nc_get_att_double(ncid,var_id,"missing_value",&missing_value);
  if(rc)
    rc=nc_get_att_double(ncid,var_id,"_FillValue",&missing_value);
  if(rc)
  {
    fprintf(stderr,"WARNING402: Cannot read missing for fill value in '%s': %s, set to %g.\n",
            argv[i],nc_strerror(rc),netcdf_config.missing_value.f);
    missing_value=netcdf_config.missing_value.f;
  }
  out=fopen(argv[i+1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[i+1],strerror(errno));
    free(dimids);
    free(lon);
    free(lat);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  header.cellsize_lon=(float)((lon[lon_len-1]-lon[0])/(lon_len-1));
  header.cellsize_lat=(float)fabs((lat[lat_len-1]-lat[0])/(lat_len-1));
  if(!israw)
    fwriteheader(out,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  header.ncell=0;
  if(latlon)
  {
    index1=first+1;
    len1=lon_len;
    index2=first;
    len2=lat_len;
  }
  else
  {
    index1=first;
    len1=lat_len;
    index2=first+1;
    len2=lon_len;
  }
  for(offsets[index1]=0;offsets[index1]<len1;offsets[index1]++)
  {
    for(offsets[index2]=0;offsets[index2]<len2;offsets[index2]++)
    {
      rc=nc_get_vara_double(ncid,var_id,offsets,counts,&data);
      error(rc);
      if((!isnan(missing_value) && !isnan(data) && data!=missing_value) ||
          (isnan(missing_value) && !isnan(data)))
      {
        switch(header.datatype)
        {
          case LPJ_FLOAT:
            coord_f.lat=(float)lat[offsets[first]];
            coord_f.lon=(float)lon[offsets[first+1]];
            fwrite(&coord_f,sizeof(coord_f),1,out);
            break;
          case LPJ_DOUBLE:
            coord_d.lat=lat[offsets[first]];
            coord_d.lon=lon[offsets[first+1]];
            fwrite(&coord_d,sizeof(coord_d),1,out);
            break;
          default:
            coord.lat=(short)(lat[offsets[first]]/header.scalar);
            coord.lon=(short)(lon[offsets[first+1]]/header.scalar);
#ifdef DEBUG
            printf("%.3f %3f %d %d\n",lat[offsets[1]],lon[offsets[2]],coord.lat,coord.lon);
#endif
            fwrite(&coord,sizeof(coord),1,out);
        }
        header.ncell++;
      }
    }
  }
  header.firstcell=0;
  header.nyear=1;
  header.nstep=1;
  header.timestep=1;
  header.firstyear=1901;
  header.nbands=2;
  header.order=CELLYEAR;
  if(!israw)
  {
    rewind(out);
    fwriteheader(out,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  }
  fclose(out);
  free(dimids);
  free(lon);
  free(lat);
  nc_close(ncid);
  printf("Number of cells: %d\n",header.ncell);
  if(isjson)
  {
    out_json=malloc(strlen(argv[i+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[i+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    fprintjson(out,argv[i+1],NULL,argv[0],NULL,arglist,&header,NULL,NULL,NULL,0,"grid","degree",NULL,"cell coordinates",NULL,LPJ_SHORT,(israw) ? RAW : CLM,LPJGRID_HEADER,FALSE,LPJGRID_VERSION);
    fclose(out);
  }
  return EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
